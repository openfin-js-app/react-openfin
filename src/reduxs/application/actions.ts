import { createAction, ActionFunctionAny, Action } from 'redux-actions';
import {WindowOptions} from 'redux-openfin';

import { IApplicationNewSnackbarOption, IApplicationCloseSnackBarOption } from './types';
import * as React from "react";

export const APPLICATION_SET_LOADING_MSG:string = 'APPLICATION_SET_LOADING_MSG';
export const APPLICATION_STARTED:string = 'APPLICATION_STARTED';
export const APPLICATION_CHILD_STARTED:string = 'APPLICATION_CHILD_STARTED';
export const APPLICATION_NOTIFICATION_STARTED:string = 'APPLICATION_NOTIFICATION_STARTED';
export const APPLICATION_READY:string = 'APPLICATION_READY';
export const APPLICATION_DRAWER_TOGGLE:string = 'APPLICATION_DRAWER_TOGGLE';
export const APPLICATION_TOGGLE_WINDOW_STATE:string = 'APPLICATION_TOGGLE_WINDOW_STATE';

export const applicationSetLoadingMsg:ActionFunctionAny<Action<{loadingMsg:string}>> =
    createAction<{loadingMsg:string},string>(APPLICATION_SET_LOADING_MSG,(loadingMsg:string)=>({loadingMsg}));
export const applicationStarted:ActionFunctionAny<Action<void>> = createAction(APPLICATION_STARTED);
export const applicationChildStarted:ActionFunctionAny<Action<void>> = createAction(APPLICATION_CHILD_STARTED);
export const applicationNotificationStarted:ActionFunctionAny<Action<void>> = createAction(APPLICATION_NOTIFICATION_STARTED);
export const applicationReady:ActionFunctionAny<Action<void>> = createAction(APPLICATION_READY);
export const applicationDrawerToggle:ActionFunctionAny<Action<void>> = createAction(APPLICATION_DRAWER_TOGGLE);
export const applicationToogleWindowState:ActionFunctionAny<Action<void>> = createAction(APPLICATION_TOGGLE_WINDOW_STATE);

// snackbar

export const APPLICATION_NEW_SNACKBAR='APPLICATION_NEW_SNACKBAR';
export const APPLICATION_SET_SNACKBAR_STATUS='APPLICATION_SET_SNACKBAR_STATUS';
export const APPLICATION_PROCESS_SNACKBAR_QUEUE='APPLICATION_PROCESS_SNACKBAR_QUEUE';
export const APPLICATION_CLOSE_SNACKBAR='APPLICATION_CLOSE_SNACKBAR';

export const applicationNewSnackbar:ActionFunctionAny<Action<IApplicationNewSnackbarOption>>
    = createAction(APPLICATION_NEW_SNACKBAR, (option:IApplicationNewSnackbarOption)=>(option));
export const applicationSetSnackbarStatus = createAction<{snackBarOpen:boolean},boolean>(APPLICATION_SET_SNACKBAR_STATUS, (snackBarOpen:boolean)=>({snackBarOpen}));
export const applicationProcessSnackbarQueue:ActionFunctionAny<Action<{}>>
    = createAction(APPLICATION_PROCESS_SNACKBAR_QUEUE);
export const applicationCloseSnackbar:ActionFunctionAny<Action<IApplicationCloseSnackBarOption>>
    = createAction(APPLICATION_CLOSE_SNACKBAR, (event: React.SyntheticEvent<any>, reason: string)=>({event,reason}));


// snap dock
export const APPLICATION_UPDATE_DOCK_STATUS = 'APPLICATION_UPDATE_DOCK_STATUS';
export const applicationUpdateDockStatus:ActionFunctionAny<Action<{docked:boolean}>>
    = createAction(APPLICATION_UPDATE_DOCK_STATUS, (docked:boolean)=>({docked}));

// launch bar
export const APPLICATION_LAUNCH_BAR_TOGGLE = 'APPLICATION_LAUNCH_BAR_TOGGLE';
export const APPLICATION_LAUNCH_BAR_TOGGLE_COLLAPSE = 'APPLICATION_LAUNCH_BAR_TOGGLE_COLLAPSE';
export const APPLICATION_LAUNCH_NEW_WINDOW = 'APPLICATION_LAUNCH_NEW_WINDOW';
export const APPLICATION_LAUNCH_BAR_CLOSE = 'APPLICATION_LAUNCH_BAR_CLOSE';
export const applicationLaunchBarToggle:ActionFunctionAny<Action<void>> = createAction(APPLICATION_LAUNCH_BAR_TOGGLE);
export const applicationLaunchBarToggleCollapse:ActionFunctionAny<Action<void>> = createAction(APPLICATION_LAUNCH_BAR_TOGGLE_COLLAPSE);
export const applicationLaunchNewWindow:ActionFunctionAny<Action<Partial<WindowOptions>>>
    = createAction(APPLICATION_LAUNCH_NEW_WINDOW,(appJson:Partial<WindowOptions>)=>(appJson));
export const applicationLaunchBarClose:ActionFunctionAny<Action<void>> = createAction(APPLICATION_LAUNCH_BAR_CLOSE);

// network
export const APPLICATION_NETWORK_ONLINE = 'APPLICATION_NETWORK_ONLINE';
export const applicationNetworkOnline:ActionFunctionAny<Action<void>> = createAction(APPLICATION_NETWORK_ONLINE);
export const APPLICATION_NETWORK_OFFLINE = 'APPLICATION_NETWORK_OFF_LINE';
export const applicationNetworkOffline:ActionFunctionAny<Action<void>> = createAction(APPLICATION_NETWORK_OFFLINE);
