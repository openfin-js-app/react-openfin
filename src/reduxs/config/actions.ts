import {createAction, ActionFunctionAny, Action} from 'redux-actions';

import {
    IConfigDexie,
    IConfigResetOption,
    IConfigUpdateOneFieldOption,
    IConfigUpdateGlobalFilterStrOption,
    IConfigDoUpdateOneFieldOption
} from './types';

export const CONFIG_RESET:string = 'CONFIG_RESET';

// redux
export const CONFIG_DO_UPDATE_ONE_FIELD:string = 'CONFIG_DO_UPDATE_ONE_FIELD';
export const CONFIG_UPDATE_GLOBAL_FILTER_STR:string = 'CONFIG_UPDATE_GLOBAL_FILTER_STR';
export const configReset:ActionFunctionAny<Action<IConfigResetOption>>
    = createAction(CONFIG_RESET, (option)=>(option));
export const configDoUpdateOneField:ActionFunctionAny<Action<IConfigDoUpdateOneFieldOption>>
    = createAction(CONFIG_DO_UPDATE_ONE_FIELD, (option)=>(option));
export const configUpdateGlobalFilterStr:ActionFunctionAny<Action<IConfigUpdateGlobalFilterStrOption>>
    = createAction(CONFIG_UPDATE_GLOBAL_FILTER_STR, (option)=>(option));

// saga
export const CONFIG_LOAD_FROM_DEXIE:string = 'CONFIG_LOAD_FROM_DEXIE';
export const configLoadFromDexie:ActionFunctionAny<Action<void>>
    = createAction(CONFIG_LOAD_FROM_DEXIE,);
export const CONFIG_DO_UPDATE_ONE_FIELD_IN_DEXIE:string = 'CONFIG_DO_UPDATE_ONE_FIELD_IN_DEXIE';
export const configDoUpdateOneFieldInDexie:ActionFunctionAny<Action<IConfigDexie>>
    = createAction(CONFIG_DO_UPDATE_ONE_FIELD_IN_DEXIE, (option)=>(option));

export const CONFIG_UPDATE_ONE_FIELD:string = 'CONFIG_UPDATE_ONE_FIELD';
export const configUpdateOneField:ActionFunctionAny<Action<IConfigUpdateOneFieldOption>>
    = createAction(CONFIG_UPDATE_ONE_FIELD, (option)=>(option));



// optional actions depending on the config application values

export const CONFIG_UPDATE_NEW_WINDOW_POSITION:string = 'CONFIG_UPDATE_NEW_WINDOW_POSITION';
export const CONFIG_UPDATE_NEW_WINDOW_POSITION_ADD_DELTA:string = 'CONFIG_UPDATE_NEW_WINDOW_POSITION_ADD_DELTA';
export const CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_TOP:string = 'CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_TOP';
export const CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_LEFT:string = 'CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_LEFT';
export const configUpdateNewWindowPosition:ActionFunctionAny<Action<void>>
    = createAction(CONFIG_UPDATE_NEW_WINDOW_POSITION);
export const configUpdateNewWindowPositionAddDelta:ActionFunctionAny<Action<void>>
    = createAction(CONFIG_UPDATE_NEW_WINDOW_POSITION_ADD_DELTA);
export const configUpdateNewWindowPositionResetTop:ActionFunctionAny<Action<void>>
    = createAction(CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_TOP);
export const configUpdateNewWindowPositionResetLeft:ActionFunctionAny<Action<void>>
    = createAction(CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_LEFT);