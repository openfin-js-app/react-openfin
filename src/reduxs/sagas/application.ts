import {Action} from "redux-actions";
import { all, call, delay, put, putResolve, race, take, takeLatest, takeEvery, fork, select, actionChannel } from 'redux-saga/effects';
import { Docking ,System, Event, Window } from 'redux-openfin';
import { GetGroupResPayload, NewWindowResPayload, WrapResPayload } from "redux-openfin/window";

import { getAllShownItems } from '../../utils/generalUtils';

import { findOneFieldVal } from '../../dexies/configDao';

import {
    // types
    IRootState,
    IReadyToClose,
    APPLICATION_LAUNCH_BAR_STATUS,

    // actions
    applicationSetLoadingMsg,
    APPLICATION_START,
    applicationAwait,
    APPLICATION_READY,
    applicationStarted,
    APPLICATION_CHILD_START,
    applicationChildAwait,
    APPLICATION_CHILD_READY,
    applicationChildStarted,
    APPLICATION_NOTIFICATION_START,
    applicationNotificationAwait,
    APPLICATION_NOTIFICATION_READY,
    applicationNotificationStarted,
    applicationCurWinClosing,
    APPLICATION_CUR_WIN_READY_TO_CLOSE,
    APPLICATION_NEW_SNACKBAR,
    APPLICATION_CLOSE_SNACKBAR,
    applicationUpdateDockStatus,
    APPLICATION_TOGGLE_WINDOW_STATE,
    applicationNewSnackbar,
    applicationSetSnackbarStatus,
    applicationProcessSnackbarQueue,
    APPLICATION_LAUNCH_BAR_TOGGLE,
    applicationLaunchBarToggled,
    APPLICATION_LAUNCH_BAR_TOGGLE_COLLAPSE,
    APPLICATION_LAUNCH_BAR_CLOSE,
    APPLICATION_LAUNCH_NEW_WINDOW,
    configLoadFromDexie,
    configUpdateNewWindowPosition,
} from '..';

import initState from '../../init';

export const LOADING_VIEW_UUID=`${initState.finUuid}-loading-view`;
let loadingWindow = null;
export const LAUNCHBAR_VIEW_UUID=`${initState.finUuid}-launchbar-view`;
let launchbarWindow = null;

const ENABLE_LOADING_VIEW=initState.config.enableLoadingView;

const LOADING_BANNER_WIDTH = initState.config.defaultLoadingBannerWidth;
const LOADING_BANNER_HEIGHT = initState.config.defaultLoadingBannerHeight;
const DEFAULT_WIDTH = initState.config.defaultAppWidth;
const DEFAULT_HEIGHT = initState.config.defaultAppHeight;

export const getSnackBarOpen = (state:IRootState) => state.application.snackBarOpen;
export const getLaunchBarCollapse = (state:IRootState) => state.application.launchBarCollapse;
export const getWindowsState = (state:IRootState)  => state.application.windowsState;
export const getNewWindowTop = (state:IRootState)  => state.config.application.newWinTop;
export const getNewWindowLeft = (state:IRootState)  => state.config.application.newWinLeft;
export const getNewWindowWidth = (state:IRootState)  => state.config.application.newWinWidth;
export const getNewWindowHeight = (state:IRootState)  => state.config.application.newWinHeight;

export function* handleShowLoadingView(monitorRect) {

    const WINDOW_WIDTH  = monitorRect.right - monitorRect.left;
    const WINDOW_HEIGHT = monitorRect.bottom - monitorRect.top;
    const _LOADING_BANNER_WIDTH     = Math.min( LOADING_BANNER_WIDTH, WINDOW_WIDTH * 0.6387 );
    const _LOADING_BANNER_HEIGHT    = Math.min( LOADING_BANNER_HEIGHT, WINDOW_HEIGHT * 0.324074 );

    const newWindowResAction:Action<NewWindowResPayload> = yield call(Window.asyncs.newWindow,Window.actions.newWindow({
        name:LOADING_VIEW_UUID,
        url:'/loading',
        frame:false,
        resizable:false,
        state:'normal',
        autoShow:true,
        defaultCentered:true,
        defaultLeft:(monitorRect.right - monitorRect.left)/2 - _LOADING_BANNER_WIDTH/2,
        defaultTop:(monitorRect.bottom - monitorRect.top)/2 - _LOADING_BANNER_HEIGHT/2,
        defaultWidth:_LOADING_BANNER_WIDTH,
        defaultHeight: _LOADING_BANNER_HEIGHT,
    }));
    loadingWindow = newWindowResAction.payload.window;

    loadingWindow.setBounds(
        (monitorRect.right - monitorRect.left)/2 - _LOADING_BANNER_WIDTH/2,
        (monitorRect.bottom - monitorRect.top)/2 - _LOADING_BANNER_HEIGHT/2,
        _LOADING_BANNER_WIDTH,
         _LOADING_BANNER_HEIGHT
    );
    loadingWindow.bringToFront();

}

export function* handleHideFromLoadingView(monitorRect, targetUrl?:string) {

    // after the sagas loaded, redirect to default page/view
    if(targetUrl && targetUrl.length > 0){
        initState.hist.push(targetUrl);
    }else if ( initState.config.defaultViewUrl && initState.config.defaultViewUrl.length > 0){
        initState.hist.push( initState.config.defaultViewUrl );
    }else{
        initState.hist.push('/dashboard/view-one');
    }

    yield delay(200);

    if (loadingWindow){
        loadingWindow.close(true);
    }
    yield put(Window.actions.show({force:true}))
}

export function* handleApplicationLoading() {

    const dbSavedLang = yield call(findOneFieldVal,'application','language');
    if (dbSavedLang){
        initState.i18n.changeLanguage(dbSavedLang);
    }

    const currentIsLoadingView =
        (new URL(window.location.href).pathname.indexOf('loading')>-1) ||
        (new URL(window.location.href).pathname.indexOf('index.html')>-1);


    const monitorInfoAction = yield call(System.asyncs.getMonitorInfo,System.actions.getMonitorInfo({}));
    const monitorRect = monitorInfoAction.payload.primaryMonitor.monitorRect;

    if (ENABLE_LOADING_VIEW && currentIsLoadingView){
        yield* handleShowLoadingView(monitorRect) as any;
    }

    yield putResolve(applicationSetLoadingMsg('init'));

    yield all([
        putResolve(configLoadFromDexie()),
        call(System.asyncs.getMachineId,System.actions.getMachineId({})),
        // getDeviceUserId might fail, thus use flux syntax........
        putResolve(System.actions.getDeviceUserId({})),
        take(System.actions.GET_DEVICE_USER_ID_RES),
        call(System.asyncs.getEnvironmentVariable,System.actions.getEnvironmentVariable({env:'USERNAME'})),
        call(System.asyncs.getEnvironmentVariable,System.actions.getEnvironmentVariable({env:'computername'})),
        call(System.asyncs.getEnvironmentVariable,System.actions.getEnvironmentVariable({env:'HOSTNAME'})),
        call(System.asyncs.getVersion,System.actions.getVersion({})),
        call(System.asyncs.getHostSpecs,System.actions.getHostSpecs({})),
        call(Window.asyncs.getState,Window.actions.getState({})),
        // delay for loading view render, could be removed
        // delay(800),
    ]);



    yield putResolve(applicationAwait());
    const { readyRes, timeout } = yield race({
        readyRes: take(APPLICATION_READY),
        timeout:delay(initState.config.onAppAwaitDelayTime),
    });
    // console.log("[react-openfin]::app saga take APPLICATION_READY or time out", readyRes, timeout);
    const readyPayload = readyRes?readyRes.payload:{};

    yield putResolve(applicationSetLoadingMsg('ready'));
    yield putResolve(applicationStarted());

    if (ENABLE_LOADING_VIEW && currentIsLoadingView){
        yield* handleHideFromLoadingView(monitorRect, (readyPayload && readyPayload.targetUrl)?readyPayload.targetUrl:void 0) as any;
    }
    yield put(applicationSetLoadingMsg(''));

    yield call(Window.asyncs.bringToFront,Window.actions.bringToFront({}));
}

export function* handleApplicationChildLoading() {


    const dbSavedLang = yield call(findOneFieldVal,'application','language');
    if (dbSavedLang){
        initState.i18n.changeLanguage(dbSavedLang);
    }

    yield all([
        call(Window.asyncs.getBounds,Window.actions.getBounds({})),
        putResolve(configLoadFromDexie()),
    ]);

    const groupedWindowsRes = yield call(Window.asyncs.getGroup,Window.actions.getGroup({}));
    const groupedWindows = (groupedWindowsRes.payload as GetGroupResPayload).windows;
    if(groupedWindows && groupedWindows.length > 0){
        yield put(applicationUpdateDockStatus(true));
    }else{
        yield put(applicationUpdateDockStatus(false));
    }

    yield putResolve(applicationChildAwait());
    const { readyRes, timeout } = yield race({
        readyRes: take(APPLICATION_CHILD_READY),
        timeout:delay(initState.config.onAppChildAwaitDelayTime),
    });
    // console.log("[react-openfin]::app saga take APPLICATION_CHILD_READY or time out", readyRes, timeout);
    const readyPayload = readyRes?readyRes.payload:{};
    if(readyPayload && readyPayload.targetUrl){
        initState.hist.push(readyPayload.targetUrl);
    }
    yield putResolve(applicationChildStarted());

}

export function* handleApplicationNotificationLoading() {
    const dbSavedLang = yield call(findOneFieldVal,'application','language');
    if (dbSavedLang){
        initState.i18n.changeLanguage(dbSavedLang);
    }

    yield all([
        call(Window.asyncs.getBounds,Window.actions.getBounds({})),
        putResolve(configLoadFromDexie()),
    ]);


    yield putResolve(applicationNotificationAwait());
    const { readyRes, timeout } = yield race({
        readyRes: take(APPLICATION_NOTIFICATION_READY),
        timeout:delay(initState.config.onAppNotificationAwaitDelayTime),
    });
    // console.log("[react-openfin]::app saga take APPLICATION_NOTIFICATION_READY or time out", readyRes, timeout);
    const readyPayload = readyRes?readyRes.payload:{};
    if(readyPayload && readyPayload.targetUrl){
        initState.hist.push(readyPayload.targetUrl);
    }
    yield putResolve(applicationNotificationStarted());

}

export function* handleApplicationExit() {

    yield putResolve(applicationCurWinClosing());
    const { readyToClose, timeout } = yield race({
        readyToClose : take(APPLICATION_CUR_WIN_READY_TO_CLOSE),
        timeout : delay(initState.config.onAppClosingAwaitDelayTime),
    });
    // console.log("[react-openfin]::app saga take APPLICATION_CUR_WIN_READY_TO_CLOSE or time out", readyToClose, timeout);

    let forceToClose:boolean = true;

    if (readyToClose){
        console.log("[react-openfin]::app saga client response and ready to close");
        const payload:IReadyToClose = readyToClose.payload;
        if (payload.skipClosing){
            console.log("[react-openfin]::app saga client response to skip closing");
            forceToClose = false;
        }
    }

    if (forceToClose){
        yield putResolve(Window.actions.close({force:true}));
    }

}

export function* handleApplicationAddNewSnackBar() {
    const snackBarOpen = yield select(getSnackBarOpen);
    if(snackBarOpen){
        yield put(applicationSetSnackbarStatus(false));
    }else{
        yield put(applicationProcessSnackbarQueue());
    }
}

export function* handleApplicationCloseSnackBar(action) {
    const {event, reason} = action.payload;
    if(reason!=='clickaway'){
        return;
    }else{
        yield put(applicationSetSnackbarStatus(false));
    }
}

export function* handleToggleWindowState(){
    const windowState = yield select(getWindowsState);
    if (windowState === 'maximized'){
        yield call(Window.asyncs.restore,Window.actions.restore({}));
    }else if (windowState === 'normal'){
        yield call(Window.asyncs.maximize,Window.actions.maximize({}));
    }
}

export function* handleApplicationLaunchBarToggle(){

    const launchBarCollapse = yield select(getLaunchBarCollapse);
    const getBoundsAction = yield call(Window.asyncs.getBounds,Window.actions.getBounds({}));
    const getBoundsActionPayload = getBoundsAction.payload;

    const mainWindowAction:Action<WrapResPayload> = yield call(Window.asyncs.wrap,Window.actions.wrap({
        appUuid: initState.finUuid,
        windowName: initState.finUuid,
    }));
    const mainWindow = mainWindowAction.payload.window;

    const launchbarWindowAction:Action<WrapResPayload> = yield call(Window.asyncs.wrap,Window.actions.wrap({
        appUuid: initState.finUuid,
        windowName: LAUNCHBAR_VIEW_UUID,
    }));


    if (launchbarWindowAction.payload.window.nativeWindow){
        launchbarWindow = launchbarWindowAction.payload.window;
        mainWindow.show(true);
        launchbarWindow.close();
        yield put(applicationLaunchBarToggled(APPLICATION_LAUNCH_BAR_STATUS.SWITCH_TO_MAIN_WIN));
    }else{
        launchbarWindow = null;
        const newWindowResAction:Action<NewWindowResPayload> = yield call(Window.asyncs.newWindow,Window.actions.newWindow({
            name:LAUNCHBAR_VIEW_UUID,
            url:'/launchBar',
            frame:false,
            resizable:false,
            state:'normal',
            autoShow:true,
            defaultLeft:getBoundsActionPayload.left,
            defaultTop:getBoundsActionPayload.top,
            defaultWidth:getAllShownItems(initState.launchBarItems).length<10? getAllShownItems(initState.launchBarItems).length*64+88:664,
            defaultHeight: 64,
            minWidth:88,
            minHeight:64,
        }));
        launchbarWindow = newWindowResAction.payload.window;

        if (launchBarCollapse){
            launchbarWindow.setBounds(
                getBoundsActionPayload.left,
                getBoundsActionPayload.top,
                88,
                64,
            );
        }else{
            launchbarWindow.setBounds(
                getBoundsActionPayload.left,
                getBoundsActionPayload.top,
                getAllShownItems(initState.launchBarItems).length<10? getAllShownItems(initState.launchBarItems).length*64+88:664,
                64,
            );
        }

        launchbarWindow.bringToFront();
        mainWindow.hide();
        yield put(applicationLaunchBarToggled(APPLICATION_LAUNCH_BAR_STATUS.SWITCH_TO_LAUNCHBAR));
    }

}

export function* handleApplicationLaunchBarToggleCollapse() {
    const launchBarCollapse = yield select(getLaunchBarCollapse);

    const getBoundsAction = yield call(Window.asyncs.getBounds,Window.actions.getBounds({}));
    const getBoundsActionPayload = getBoundsAction.payload;

    if (launchBarCollapse){
        yield call(Window.asyncs.setBounds,Window.actions.setBounds({
            left:getBoundsActionPayload.left,
            top:getBoundsActionPayload.top,
            width:88,
            height:64,
        }));
    }else{
        yield call(Window.asyncs.setBounds,Window.actions.setBounds({
            left:getBoundsActionPayload.left,
            top:getBoundsActionPayload.top,
            width:getAllShownItems(initState.launchBarItems).length<10? getAllShownItems(initState.launchBarItems).length*64+88:664,
            height:64,
        }));
    }
}

export function* handleApplicationLaunchBarClose() {
    const mainWindowAction:Action<WrapResPayload> = yield call(Window.asyncs.wrap,Window.actions.wrap({
        appUuid: initState.finUuid,
        windowName: initState.finUuid,
    }));
    const mainWindow = mainWindowAction.payload.window;
    mainWindow.close(false);
}

export function* handleApplicationLaunchNewWindow(action) {

    if (window.name === initState.finUuid){

        const appJson = action.payload;
        const windowName = appJson.name;

        const wrapWindowAction:Action<WrapResPayload> = yield call(Window.asyncs.wrap,Window.actions.wrap({
            appUuid: initState.finUuid,
            windowName,
        }));

        if (
            wrapWindowAction.payload &&
            wrapWindowAction.payload.window &&
            wrapWindowAction.payload.window.nativeWindow
        ){
            // already created, not need to create anymore
            const theWindow = wrapWindowAction.payload.window;
            theWindow.show(true);
            theWindow.bringToFront();
        }else{
            // not created, need to create one
            const defaultWidth = yield select(getNewWindowWidth);
            const defaultHeight = yield select(getNewWindowHeight);
            const defaultTop = yield select(getNewWindowTop);
            const defaultLeft = yield select(getNewWindowLeft);

            if(!appJson.defaultWidth){ appJson.defaultWidth = defaultWidth}
            if(!appJson.defaultHeight){ appJson.defaultHeight = defaultHeight}
            if(!appJson.defaultTop){ appJson.defaultTop = defaultTop}
            if(!appJson.defaultLeft){ appJson.defaultLeft = defaultLeft}

            const newWindowResAction:Action<NewWindowResPayload> = yield call(Window.asyncs.newWindow,Window.actions.newWindow(appJson));
            const newWindow = newWindowResAction.payload.window;
            newWindow.bringToFront();

            yield put(configUpdateNewWindowPosition());
        }

    }

}

export function* handleGroupChanged(action){
    const {
        sourceWindowName, targetWindowName, memeberOf, reason
    } = action.payload;


    if (reason === Docking.types.GroupEventReason.JOIN){
        if(sourceWindowName === window.name){
            yield put(applicationNewSnackbar({
                message:'Joined group',
                variant:'primary'
            }))
        }else if (targetWindowName === window.name){
            yield put(applicationNewSnackbar({
                message:'Been joined',
                variant:'rose'
            }))
        }
    }else if (
        reason === Docking.types.GroupEventReason.LEAVE &&
        sourceWindowName === window.name
    ){
        yield put(applicationNewSnackbar({
            message:'Left group',
            variant:'primary'
        }))
    }else if (
        reason === Docking.types.GroupEventReason.DISBAND &&
        sourceWindowName === window.name
    ){
        yield put(applicationNewSnackbar({
            message:'Got disbanded',
            variant:'rose'
        }))
    }

}

export default function* (){
    yield takeLatest(APPLICATION_START,handleApplicationLoading);
    yield takeLatest(APPLICATION_CHILD_START,handleApplicationChildLoading);
    yield takeLatest(APPLICATION_NOTIFICATION_START,handleApplicationNotificationLoading);
    yield takeLatest(Event.actionDicts.windowEventDictByName['close-requested'].type,handleApplicationExit);
    yield takeLatest(APPLICATION_TOGGLE_WINDOW_STATE,handleToggleWindowState);
    yield takeLatest(APPLICATION_NEW_SNACKBAR,handleApplicationAddNewSnackBar);
    yield takeLatest(APPLICATION_CLOSE_SNACKBAR,handleApplicationCloseSnackBar);
    yield takeLatest(APPLICATION_LAUNCH_BAR_TOGGLE,handleApplicationLaunchBarToggle);
    yield takeLatest(APPLICATION_LAUNCH_BAR_TOGGLE_COLLAPSE,handleApplicationLaunchBarToggleCollapse);
    yield takeLatest(APPLICATION_LAUNCH_BAR_CLOSE,handleApplicationLaunchBarClose);
    yield takeLatest(APPLICATION_LAUNCH_NEW_WINDOW,handleApplicationLaunchNewWindow);
    yield takeEvery(Event.actionDicts.windowEventDictByName['group-changed'].type,handleGroupChanged);
}