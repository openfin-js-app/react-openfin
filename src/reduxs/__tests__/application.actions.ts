import * as Actions from '../application/actions';

describe('Application actions',()=>{

    it('applicationSetLoadingMsg action',()=>{
        expect(Actions.applicationSetLoadingMsg('loadingMsg')).toMatchSnapshot();
    })

})
