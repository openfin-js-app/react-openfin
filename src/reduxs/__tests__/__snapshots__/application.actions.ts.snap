// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application actions applicationAwait action 1`] = `
Object {
  "type": "REACT_OPENFIN_RES::APPLICATION_AWAIT",
}
`;

exports[`Application actions applicationChildAwait action 1`] = `
Object {
  "type": "REACT_OPENFIN_RES::APPLICATION_CHILD_AWAIT",
}
`;

exports[`Application actions applicationChildReady action 1`] = `
Object {
  "payload": Object {
    "targetUrl": "targetUrl",
  },
  "type": "REACT_OPENFIN_REQ::APPLICATION_CHILD_READY",
}
`;

exports[`Application actions applicationChildStart action 1`] = `
Object {
  "type": "REACT_OPENFIN_REQ::APPLICATION_CHILD_START",
}
`;

exports[`Application actions applicationChildStarted action 1`] = `
Object {
  "type": "REACT_OPENFIN_RES::APPLICATION_CHILD_STARTED",
}
`;

exports[`Application actions applicationCloseSnackbar action 1`] = `
Object {
  "payload": Object {
    "event": Object {},
    "reason": "reason",
  },
  "type": "REACT_OPENFIN_REQ::APPLICATION_CLOSE_SNACKBAR",
}
`;

exports[`Application actions applicationCurWinClosing action 1`] = `
Object {
  "type": "REACT_OPENFIN_RES::APPLICATION_CUR_WIN_CLOSING",
}
`;

exports[`Application actions applicationCurWinReadyToClose action 1`] = `
Object {
  "payload": Object {},
  "type": "REACT_OPENFIN_REQ::APPLICATION_CUR_WIN_READY_TO_CLOSE",
}
`;

exports[`Application actions applicationDrawerToggle action 1`] = `
Object {
  "type": "REACT_OPENFIN_REQ::APPLICATION_DRAWER_TOGGLE",
}
`;

exports[`Application actions applicationLaunchBarClose action 1`] = `
Object {
  "type": "REACT_OPENFIN_REQ::APPLICATION_LAUNCH_BAR_CLOSE",
}
`;

exports[`Application actions applicationLaunchBarToggle action 1`] = `
Object {
  "type": "REACT_OPENFIN_REQ::APPLICATION_LAUNCH_BAR_TOGGLE",
}
`;

exports[`Application actions applicationLaunchBarToggleCollapse action 1`] = `
Object {
  "type": "REACT_OPENFIN_REQ::APPLICATION_LAUNCH_BAR_TOGGLE_COLLAPSE",
}
`;

exports[`Application actions applicationLaunchBarToggled action 1`] = `
Object {
  "payload": Object {
    "status": 1,
  },
  "type": "REACT_OPENFIN_RES::APPLICATION_LAUNCH_BAR_TOGGLED",
}
`;

exports[`Application actions applicationLaunchBarToggled action 2`] = `
Object {
  "payload": Object {
    "status": 2,
  },
  "type": "REACT_OPENFIN_RES::APPLICATION_LAUNCH_BAR_TOGGLED",
}
`;

exports[`Application actions applicationLaunchNewWindow action 1`] = `
Object {
  "payload": Object {},
  "type": "REACT_OPENFIN_REQ::APPLICATION_LAUNCH_NEW_WINDOW",
}
`;

exports[`Application actions applicationNetworkOffline action 1`] = `
Object {
  "type": "REACT_OPENFIN::APPLICATION_NETWORK_OFF_LINE",
}
`;

exports[`Application actions applicationNetworkOnline action 1`] = `
Object {
  "type": "REACT_OPENFIN::APPLICATION_NETWORK_ONLINE",
}
`;

exports[`Application actions applicationNewSnackbar action 1`] = `
Object {
  "payload": Object {
    "message": "message",
    "variant": "primary",
  },
  "type": "REACT_OPENFIN_REQ::APPLICATION_NEW_SNACKBAR",
}
`;

exports[`Application actions applicationNotificationAwait action 1`] = `
Object {
  "type": "REACT_OPENFIN_RES::APPLICATION_NOTIFICATION_AWAIT",
}
`;

exports[`Application actions applicationNotificationReady action 1`] = `
Object {
  "payload": Object {
    "targetUrl": "targetUrl",
  },
  "type": "REACT_OPENFIN_REQ::APPLICATION_NOTIFICATION_READY",
}
`;

exports[`Application actions applicationNotificationStart action 1`] = `
Object {
  "type": "REACT_OPENFIN_REQ::APPLICATION_NOTIFICATION_START",
}
`;

exports[`Application actions applicationNotificationStarted action 1`] = `
Object {
  "type": "REACT_OPENFIN_RES::APPLICATION_NOTIFICATION_STARTED",
}
`;

exports[`Application actions applicationProcessSnackbarQueue action 1`] = `
Object {
  "type": "REACT_OPENFIN_REQ::APPLICATION_PROCESS_SNACKBAR_QUEUE",
}
`;

exports[`Application actions applicationReady action 1`] = `
Object {
  "payload": Object {
    "targetUrl": "targetUrl",
  },
  "type": "REACT_OPENFIN_REQ::APPLICATION_READY",
}
`;

exports[`Application actions applicationSetLoadingMsg action 1`] = `
Object {
  "payload": Object {
    "loadingMsg": "loadingMsg",
  },
  "type": "REACT_OPENFIN_REQ::APPLICATION_SET_LOADING_MSG",
}
`;

exports[`Application actions applicationSetSnackbarStatus action 1`] = `
Object {
  "payload": Object {
    "snackBarOpen": true,
  },
  "type": "REACT_OPENFIN_REQ::APPLICATION_SET_SNACKBAR_STATUS",
}
`;

exports[`Application actions applicationStart action 1`] = `
Object {
  "type": "REACT_OPENFIN_REQ::APPLICATION_START",
}
`;

exports[`Application actions applicationStarted action 1`] = `
Object {
  "type": "REACT_OPENFIN_RES::APPLICATION_STARTED",
}
`;

exports[`Application actions applicationToggleWindowState action 1`] = `
Object {
  "type": "REACT_OPENFIN_REQ::APPLICATION_TOGGLE_WINDOW_STATE",
}
`;

exports[`Application actions applicationUpdateDockStatus action 1`] = `
Object {
  "payload": Object {
    "docked": true,
  },
  "type": "REACT_OPENFIN::APPLICATION_UPDATE_DOCK_STATUS",
}
`;
